src/ds/d500/d500-fw-update-device.cpp

```diff
@@ -37,17 +37,65 @@ ds_d500_update_device::ds_d500_update_device( std::shared_ptr< const device_info
-        // On D500 devices after transferring the FW image the internal DFU progress start on the device
-        float iteration_sleep_time_ms = (static_cast<float>(D500_FW_DFU_TIME) / 100.0f) * 1000.0f;
-        for(int i = 1; i <= 100; i++)
-        {
+        update_device::update( fw_image, fw_image_size, update_progress_callback );
+    }
+
+    bool ds_d500_update_device::wait_for_manifest_completion(std::shared_ptr<platform::usb_messenger> messenger, const rs2_dfu_state state,
+        std::chrono::seconds timeout_seconds, rs2_update_progress_callback_sptr update_progress_callback) const
+    {
+        std::chrono::seconds elapsed_seconds;
+        auto start = std::chrono::system_clock::now();
+        rs2_dfu_state dfu_state = RS2_DFU_STATE_APP_IDLE;
+        dfu_status_payload status;
+        int percentage_of_transfer = 0;
+        int iteration = 0;
+        // progress should start increase even in the 2nd iteration, 
+        // when this DFU progress is enabled by FW
+        int max_iteration_number_for_progress_start = 10;
+
+        do {
+            uint32_t transferred = 0;
+            auto sts = messenger->control_transfer(0xa1 /*DFU_GETSTATUS_PACKET*/, RS2_DFU_GET_STATUS, 0, 0, (uint8_t*)&status, sizeof(status), transferred, 5000);
+            dfu_state = status.get_state();
+            percentage_of_transfer = static_cast<int>(status.iString);
+
+            // the below code avoids process stuck when using a d5XX device, 
+            // which has a fw version without the DFU progress feature
+            if (percentage_of_transfer == 0 && 
+                ++iteration == max_iteration_number_for_progress_start)
+                return true;
+
+            std::stringstream ss;
+            ss << "DFU_GETSTATUS called, state is: " << to_string(dfu_state); 
+            ss << ", iString equals: " << percentage_of_transfer << ", and bwPollTimeOut equals: " << status.bwPollTimeout << std::endl;
+            LOG_DEBUG(ss.str().c_str());
+
             if (update_progress_callback)
-                update_progress_callback->on_update_progress( i / 100.f );
-            std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(iteration_sleep_time_ms)));
-        }
+            {
+                auto progress_for_bar = compute_progress(static_cast<float>(percentage_of_transfer), 20.f, 100.f, 5.f) / 100.f;
+                update_progress_callback->on_update_progress(progress_for_bar);
+            }
+
+            if (sts != platform::RS2_USB_STATUS_SUCCESS)
+                LOG_DEBUG("control xfer error: " << platform::usb_status_to_string[sts]);
+
+            //test for dfu error state
+            if (status.is_error_state()) {
+                return false;
+            }
+
+            // FW doesn't set the bwPollTimeout value, therefore it is wrong to use status.bwPollTimeout
+            std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
This will cause a 100 seconds increase of the process no?
Why 1 sec?

Reply:
no - no 100 seconds here.
I just set more time between the polling of the progress percentage.
Approximatively:
with smaller time, the polling of the percentage would be: 1, 2, 3...
with 1 sec, it would be: 1, 5, 10...